### PR TITLE
Structural code search blog post

### PR DESCRIPTION
**Pull request by rvantonder**

### CODE REVIEW

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+  search to be available for _your_ repository, and not just the really popular
+  ones. [+1 this feature on GitHub](https://github.com/sourcegraph/sourcegraph/issues/8164)
+- **Regular expression support in structural holes.** We want to add support for
+  regexp syntax inside holes for refine search, like `\d+` to match only
+  numerical digits. [+1 this feature on GitHub](https://github.com/sourcegraph/sourcegraph/issues/8165)
+- **A richer query language.** There are ways to refine structural search with [rules](https://comby.dev/#advanced-usage) allowing richer queries. To get a taste of rules, have a look at this [CactusCon talk](https://www.youtube.com/watch?v=yOZQsZs35FA) (subtitles recommended if audio is hard to follow). If you have a use case for this and want support sooner, [+1 this feature on GitHub](https://github.com/sourcegraph/sourcegraph/issues/8166).
+- **Make it faster.** Structural search is typically slower than our regexp
+  search because it does more work. If you find it valuable, we
+  want to make it faster. [+1 this feature on GitHub](https://github.com/sourcegraph/sourcegraph/issues/8167)
```
Comment (rvantonder) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: These work items aren't planned any time soon. The main reason this is here is that I'd like to open up some way to gather a validation signal if people find certain features useful or missing.

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
+---
+title: Structural code search. At scale and from your browser.
+author: Rijnard van Tonder
+authorUrl: https://twitter.com/rvtond
+publishDate: 2020-01-24T10:00-07:00
+tags: [blog]
+slug: structural-code-search-at-scale-from-your-browser
+heroImage: https://TODO
```
Comment (rvantonder) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: FIXME

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
+---
+title: Structural code search. At scale and from your browser.
+author: Rijnard van Tonder
+authorUrl: https://twitter.com/rvtond
+publishDate: 2020-01-24T10:00-07:00
```
Comment (rvantonder) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: FIXME

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+copy_from_user(&txc.tick, &txc_p->tick, sizeof(struct timex32) -
+			   offsetof(struct timex32, tick))
+```
+
+where the argument spans multiple lines. By default, `:[hole]` syntax matches
+across multiple lines just like code structures can. An interesting thing about
+the call above is that it calculates the size of memory using `sizeof(...) -
+...`.  Calculating and checking the size of memory to copy can be more
+error-prone than simple or static values. So, one thing we could check is
+whether other calls that calulate the size of memory in a similar way to the
```
Comment (uwedeportivo) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]:
```suggestion
whether other calls that calculate the size of memory in a similar way to the
```

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
+---
+title: Structural code search. At scale and from your browser.
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]:
```suggestion
title: Going beyond regular expressions with structural code search
```

Mentioning regular expressions sets the context and by providing a comparison, entices them to see how it offers new possibilities.

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+author: Rijnard van Tonder
+authorUrl: https://twitter.com/rvtond
+publishDate: 2020-01-24T10:00-07:00
+tags: [blog]
+slug: structural-code-search-at-scale-from-your-browser
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: If you agree with the title change, you may want to change the slug as well.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 2 above)
Comment (ryan-blunden) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: You could this for `heroImage`:

Screen Shot 2020-02-01 at 10 12 44 AM

And this for the hero image at the top of the post.

Screen Shot 2020-02-01 at 10 11 16 AM

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+heroImage: https://TODO
+published: true
+---
+
+We're introducing a richer way to search code at Sourcegraph with structural
+code search. And we're making it available at scale.
+
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: Adding a section about why literal and reg expression searches with an example will help clarify why structural search is needed before jumping into what it is.

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+## What _is_ structural code search?
+
+Structural code search is the idea that you can search for _syntactic
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: Text formatting preference is to have each paragraph on a single line and use soft line wrapping in your editor but up to you.

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+author: Rijnard van Tonder
+authorUrl: https://twitter.com/rvtond
+publishDate: 2020-01-24T10:00-07:00
+tags: [blog]
+slug: structural-code-search-at-scale-from-your-browser
+heroImage: https://TODO
+published: true
+---
+
+We're introducing a richer way to search code at Sourcegraph with structural
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]: It's not clear what is meant by *richer*? Alternatives could be "enhanced", "more specialized", "next generation".

File: blogposts/structural-code-search-at-scale-from-your-browser.md [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+code search. And we're making it available at scale.
```
Comment (ryan-blunden) [inline blogposts/structural-code-search-at-scale-from-your-browser.md]:
```suggestion
code search.
```

To me, this puts into question whether everything we do is at scale so I'd remove it.

**Comment by tsenart** [comment]
Seems like we checked-in some `.swp` files.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 5 above)
Comment (rvantonder) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: I'm a bit torn about this. I think the 'going beyond regex' is great. On the other hand, there exist tools to do similar things locally, or you can even just run comby locally. Why then bother using Sourcegraph? IMO the problem that Sourcegraph solves is scale (it works large scale, unlike existing tools, and there is some technically interesting parts on this) and convenience (from your browser, on all repos/supported code hosts) and that's important context for structural search.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 10 above)
Comment (rvantonder) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: As above comment, I think the scale part is important from the direction of 'this is what Sourcegraph enables'. But maybe that's not clear from this sentence so I'm OK to remove it.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 9 above)
Comment (rvantonder) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: Richer in terms of expressivity. But "more expressive" isn't the right phrase. In general "more X" prompts 'more X than what?' and if X is regex then I'm not sure isn't an accurate description (the comparison is vague). Saying "new" might be the best alternative.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 5 above)
Comment (rvantonder) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: The benefit/answer to "Why bother with Sourcegraph?" might be implicit, and if we're OK with that then I'm OK with the title change.

**Comment by rvantonder** [comment]
Status on this: I plan to merge by end of week, still have kinks in the implementation I want to iron out.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 2 above)
Comment (ryan-blunden) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: @rvantonder Do you want me to add these images to the post?

Comment (rvantonder) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: Appreciate the offer! I like the idea of using a screenie of search results. It'd be cool if we could get a screenie with multiple line matches, so I think I can take care of this.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 5 above)
Comment (ryan-blunden) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: > IMO the problem that Sourcegraph solves is scale

If you also published this blog post externally, such as Medium.com, then your original title makes more sense, but as it's on Sourcegraph's blog, I don't think it's required.

What about leading with a paragraph explaining what Sourcegraph is like our release blog posts., something like:

> Sourcegraph's **Universal Code Search** is used by developers daily at companies such as Uber, Lyft, and Yelp. It works at massive scale, supporting tens of thousands of repositories, all popular languages, with powerful search and filters beyond the capabilities of IDEs, code hosts, and CLI tools.

File: blogposts/structural-code-search-at-scale-from-your-browser.md (hunk 9 above)
Comment (ryan-blunden) [reply blogposts/structural-code-search-at-scale-from-your-browser.md]: I like new the most.

**Comment by rvantonder** [comment]
I'll merge this first thing Monday. There's a nasty bug in the search code that is hurting usability and I want that fixed first. https://github.com/sourcegraph/sourcegraph/issues/7596#issuecomment-583683007

**Comment by rvantonder** [comment]
@ryan-blunden gonna merge this tomorrow. I tried some hero images with code snippets, but they all look too small in the thumbnail to look cool. So I'm rolling with a blown-up image of the structural search icon (light mode). WDYT?

Screen Shot 2020-02-11 at 11 00 45 PM

File: blogposts/going-beyond-regular-expressions-with-structural-code-search.md [hunk 11]
```diff
@@ -0,0 +1,322 @@
...
+Let's look for things in the Linux kernel. After all, why not show
+structural search working on one of the largest and most popular projects in
+open source software?
+
+One important function is `copy_from_user`, which copies content from userspace
+memory into the kernelspace memory. This function has a history of [careful
+auditing](https://www.defcon.org/images/defcon-19/dc-19-presentations/Cook/DEFCON-19-Cook-Kernel-Exploitation.pdf)
+because incorrect uses can (and have) lead to vulnerabilities. We can find all
+`copy_from_user` calls with a query like `copy_from_user(:[args])`. Try it
+live (the <svg class="mdi-icon " style="border:1px solid #2f9cf1; border-radius: 2px; fill:#2b2b2b; background:#cbd4e2" width="24" height="24" viewBox="0 0 24 24"><path d="M15,4V6H18V18H15V20H20V4M4,4V20H9V18H6V6H9V4H4Z"></path></svg> toggle means structural search is active):
```
Comment (rvantonder) [inline blogposts/going-beyond-regular-expressions-with-structural-code-search.md]: SVG here.

**Pull request merged**